File: src/main/java/com/designpatterns/dev/behavioral/chainofresponsibility/java/Handler.java
package com.designpatterns.dev.behavioral.chainofresponsibility.java;

import java.util.Stack;

public abstract class Handler {
    private Handler nextHandler;
    public Handler(Handler nextHandler) {
        this.nextHandler = nextHandler;
    }

    public void handle(Request request){
        if(doHandle(request)){ return;}
        else{
            if(nextHandler != null){
                nextHandler.handle(request);
            }
        }
    }

    public abstract boolean doHandle(Request request);
}

File: src/main/java/com/designpatterns/dev/behavioral/chainofresponsibility/kotlin/Handler.kt
package com.designpatterns.dev.behavioral.chainofresponsibility.kotlin

abstract class Handler(private var nextHandler: Handler?) {


    fun handle(request: Request){
        if(doHandle(request))return
        if(nextHandler != null){
            nextHandler!!.doHandle(request)
        }

    }

    abstract fun  doHandle(request: Request) : Boolean
}